code/globalincs/undosys.h

```diff
@@ -0,0 +1,264 @@
+ *
+ * @sa Undo_item_base
+ */
+template<typename T> class Undo_item : public Undo_item_base
+{
+public:
+	Undo_item()
+		: dest(nullptr), data(nullptr) {};
+
+	Undo_item(T& _data) {
+		save(_data);
+	};
+
+	Undo_item(T& _data, T* _cont) {
+		save(_data, _cont);
+	};
+
+	~Undo_item() {
+		if (data != nullptr) {
+			delete data;
+		}
+	};
+
+	/**
+	 * @brief Restores the saved data
+	 *
+	 * @returns A pair of const void* referencing the affected item and its container (if it was provided)
+	 * @details If successful, the data at dest has been swapped. So if this was an undo item, it's now a redo item
+	 */
+	std::pair<const void*, const void*> restore() {
+		Assert((dest != nullptr) && (data != nullptr));
+
+		T copy = *static_cast<T*>(dest);
+		*static_cast<T*>(dest) = *static_cast<T*>(data);
+		*static_cast<T*>(data) = copy;
+
+		return std::pair<const void*, const void*>(dest, container);
+	};
+
+	/*!
+	 * @brief Saves data and its destination as an undo item
+	 *
+	 * @param[in] _data The data to save
+	 * @param[in] _cont The container this data is originally found in
+	 * @details The explicit definition is here to ensure type safety
+	 */
+	void save(T& _data, T* _cont = nullptr) {
+		Assert(_data != nullptr);
+
+		dest = *_data;
+		data = new T(_data);
+		container = _cont;
+	};
+
+private:
+	T* dest;        //!< Destination of the data
+	T* data;        //!< Reference to a copy of data on the heap
+	T* container;   //!< Optional reference to this item's container
+};
+
+
+/*!
+ * @brief Class which handles multiple undo operations as a single op within the Undo_system
+ */
+class Undo_stack : public Undo_item_base
+{
+public:
+	Undo_stack();
+
+	Undo_stack(size_t size);
+
+	~Undo_stack();
+
+	/*!
+	 * @brief Restores all items within the undo stack
+	 *
+	 * @details Maintains an internal flag which will reverse direction on the next call, thereby redo-ing
+	 * 
+	 * @returns A pair of references to the last item restored
+	 */
+	std::pair<const void*, const void*> restore();
+
+	/*!
+	 * @brief Saves the item onto the undo stack
+	 *
+	 * @param[in] item      The item to save
+	 * @param[in] container (Optional) The container wherein this item is located
+	 * @note Call this _before_ you do your operation on the item
+	 */
+	template<typename T>
+	size_t save(T& item, T* container = nullptr);
+
+	/*!
+	 * @brief Calls ::reserve() on the internal vector
+	 */
+	void reserve(size_t size);
+private:
+	bool reverse;   // Direction to walk the stack. forward = false, reverse = True
+
+	std::vector<Undo_item_base*> stack;
+};
+
+/*!
+ * @brief Generic Undo/Redo system. Save whatever, restore whatever!
+ *
+ * @details Currently uses a pair of deques. Would ideally use a ring container so that the actual Undo_item instances don't go anywhere.
+ */
+class Undo_system
+{
+public:
+	Undo_system();
+
+	Undo_system(uint _undos);
```

Comment:
This parameter should be a `size_t` since it probably controls how big the buffers should be?

Reply:
Originally intended to use this as a sort of hard limit to how many undo's the system could handle, although that function is handled by a user-configurable soft limit already.  Changed to size_t in next comit.